# CoderHouse - Grupo 7 - Algoritmo de Clasificación

* <b>Profesor:</b> David Romero Acosta

* <b>Tutoria:</b> Hector Ponce Schwarz

<b>Alumnos:</b>
* Juan Pascual
* Lucas Ariel Saavedra
* Bartolome Oscar Meritello
* Jose Mornaghi

#### Primero importamos las librerias que utilizaremos en el trabajo y el dataset

In [22]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

In [23]:
data = pd.read_csv("stroke_imputed.csv")
data.head()

gender   age  hypertension  heart_disease ever_married      work_type  \
0    Male  67.0             0              1          Yes        Private   
1  Female  61.0             0              0          Yes  Self-employed   
2    Male  80.0             0              1          Yes        Private   
3  Female  49.0             0              0          Yes        Private   
4  Female  79.0             1              0          Yes  Self-employed   

  Residence_type  avg_glucose_level   smoking_status  stroke        bmi  
0          Urban             228.69  formerly smoked       1  36.600000  
1          Rural             202.21     never smoked       1  28.893237  
2          Rural             105.92     never smoked       1  32.500000  
3          Urban             171.23           smokes       1  34.400000  
4          Rural             174.12     never smoked       1  24.000000

#### Antes de comenzar con el modelo debemos hacer algunos ajustes en el dataset, convertiremos las variables categoricas en variables "dummies" 

In [24]:
data["gender"].replace({"Male": 1, "Female": 0}, inplace=True)
data["Residence_type"].replace({"Rural": 0, "Urban": 1}, inplace=True)
data["ever_married"].replace({"Yes": 1, "No": 0}, inplace=True)

data = pd.get_dummies(data, columns=['smoking_status'])
data = data.drop("smoking_status_Unknown", axis = 1)
data = pd.get_dummies(data, columns=['work_type'])
data = data.drop("work_type_Never_worked", axis = 1)

In [25]:
data.head()

gender   age  hypertension  heart_disease  ever_married  Residence_type  \
0       1  67.0             0              1             1               1   
1       0  61.0             0              0             1               0   
2       1  80.0             0              1             1               0   
3       0  49.0             0              0             1               1   
4       0  79.0             1              0             1               0   

   avg_glucose_level  stroke        bmi  smoking_status_formerly smoked  \
0             228.69       1  36.600000                               1   
1             202.21       1  28.893237                               0   
2             105.92       1  32.500000                               0   
3             171.23       1  34.400000                               0   
4             174.12       1  24.000000                               0   

   smoking_status_never smoked  smoking_status_smokes  work_type_Govt_job  \
0                            0                      0                   0   
1                            1                      0                   0   
2                            1                      0                   0   
3                            0                      1                   0   
4                            1                      0                   0   

   work_type_Private  work_type_Self-employed  work_type_children  
0                  1                        0                   0  
1                  0                        1                   0  
2                  1                        0                   0  
3                  1                        0                   0  
4                  0                        1                   0

#### Procedemos a separar el dataset de la variable objetivo

In [26]:
X = data.drop(["stroke"], axis=1)
Y = data["stroke"]

#### Ahora separamos en los sets de entrenamiento y prueba

In [27]:
(X_train, X_test,
Y_train, Y_test) = train_test_split(X,
                                    Y,
                                    stratify=Y,
                                    test_size=0.20,
                                    random_state=11)

### Vamos a utilizar dos modelos de clasificacion: KNN y RandomForest

#### Comenzando por KNearest

In [28]:
model = KNeighborsClassifier(n_neighbors=8).fit(X_train, Y_train)
model.score(X_test, Y_test)

0.9510763209393346

El modelo KNN tiene un 95.10% de aciertos sobre el set de evaluacion

#### Procedemos a utilizar el modelo RandomForest

##### Primero creamos un arbol de decisión sencillo, lo entrenamos y realizamos las predicciones

In [21]:
#Creamos un arbol de decisión sencillo y lo fiteamos
tree = DecisionTreeClassifier(random_state=11)
tree.fit(X_train, Y_train)
Y_test_pred = tree.predict(X_test)
test_accuracy = accuracy_score(Y_test, Y_test_pred)
print('% de aciertos sobre el set de evaluación:',test_accuracy)

% de aciertos sobre el set de evaluación: 0.9021526418786693


##### Finalmente creamos el modelo randomForest y procedemos a entrenarlo y realizar las predicciones

In [20]:
#Creamos un random forest!
model = RandomForestClassifier(random_state=11, n_estimators=200,
                               class_weight="balanced", max_features="log2")
model.fit(X_train, Y_train)
Y_test_pred = model.predict(X_test)
test_accuracy = accuracy_score(Y_test, Y_test_pred)
print('% de aciertos sobre el set de evaluación:',test_accuracy)

% de aciertos sobre el set de evaluación: 0.9481409001956947


Ambos modelos tienen un muy buen ajuste de los datos. KNN tiene un ligeramente mejor porcentaje de aciertos (95.10%) vs Random Forest (94.81%)